lib/publiccloud/basetest.pm

```diff
@@ -103,36 +103,54 @@ sub provider_factory {
 
 sub cleanup {
     # to be overridden by tests
+    # non-destructive cleanup for test module started by post_run_hook
+    # primary for log collection | restore clean state after test module
+    # can and be called after each module run
     return 1;
 }
 
-sub _cleanup {
+sub destroy {
+    # to be overridden by tests
+    # destrictive routine in post_fail_hook | last module of test
+    return 1;
+}
+
+sub _destroy {
     my ($self) = @_;
-    die("Cleanup called twice!") if ($self->{cleanup_called});
-    $self->{cleanup_called} = 1;
+    # run module overriden destroy
+    eval { $self->destroy(); } or bmwqemu::fctwarn("self::destroy() failed -- $@");
+    # run provider destroy
+    if ($self->{run_args} && $self->{run_args}->{my_provider}) {
+        eval { $self->{run_args}->{my_provider}->destroy($self->{run_args}); } or bmwqemu::fctwarn("provider::destroy() failed -- $@");
+    }
+}
 
+sub _cleanup {
+    my ($self) = @_;
+    # run module overriden cleanup
     eval { $self->cleanup(); } or bmwqemu::fctwarn("self::cleanup() failed -- $@");
-
-    my $flags = $self->test_flags();
-    # currently we have two cases when cleanup of image will be skipped:
-    # 1. Calling module needs to have publiccloud_multi_module => 1 test flag
-    # and not have fatal => 1. Job should not have result = 'fail'
-    return if ($flags->{publiccloud_multi_module} && !($self->{result} eq 'fail' && $flags->{fatal}));
-    # 2. Job should have PUBLIC_CLOUD_NO_CLEANUP defined and job should have result = 'fail'
-    return if ($self->{result} eq 'fail' && get_var('PUBLIC_CLOUD_NO_CLEANUP_ON_FAILURE'));
```

Comment:
Isn't the variable `PUBLIC_CLOUD_NO_CLEANUP_ON_FAILURE` now obsolete? If so, we should also remove it from `variables.md`.

Reply:
why we dropping it ? how you would stop openQA from deleting VM in the cloud from now on ?

Reply:
ok I see , you just moved variable into other place .